Weekly_Work/Week1/Code_changes-git/Metrics_Root_class.py

```diff
@@ -0,0 +1,202 @@
+import matplotlib.pyplot as plt
+import pandas as pd
+
+
+# In[7]:
+
+
+class Metric():
+    
+    def __init__(self, path, since=None, until=None, source_code_exclude_list=None):
+
+        self.since = since
+        self.until = until
+        self.source_code_exclude_list = source_code_exclude_list
+        self.repo_urls = set()
+        
+        clean_commit_list = list()
+        clean_issue_list = list()
+        clean_pr_list = list()
+        
+        with open(path, 'r') as raw_data:
+            for line in raw_data:
+                line = json.loads(line)
+                self.repo_urls.add(line['origin'])
+                clean_line = dict()
+                if line['category'] == "commit":
+                    clean_line = self._clean_commit(line)
+
+                    if self._is_source_code(clean_line):
+                        clean_commit_list.append(clean_line)
+                    
+
+                elif line['category'] == "issue":
+                    if "pull_request" not in line['data']:
+                        clean_line = self._clean_issue(line)
+                        clean_issue_list.append(clean_line)
+                    else: continue
+                    
+
+                elif line['category'] == "pull_request":
+                    clean_line = self._clean_pr(line)
+                    clean_pr_list.append(clean_line)
+                        
+                        
+                self.clean_commit_df = pd.DataFrame(clean_commit_list)
+                self.clean_issue_df = pd.DataFrame(clean_issue_list)
+                self.clean_pr_df = pd.DataFrame(clean_pr_list)
+                
+            self.clean_dict = {
+                'commit': self.clean_commit_df,
+                'issue': self.clean_issue_df,
+                'pull_request': self.clean_pr_df
+            }
+       
+        if self.since:
+            for df in self.clean_dict.values():
+                df = df[df['created_date'] >= self._str_to_dt_other(self.since)]
+        else: 
+            self.since = self._get_date("since")
+            
+        if self.until:
+            for df in self.clean_dict.values():
+                df = df[df['created_date'] < self._str_to_dt_other(self.until)]
+        else: 
+            self.until = self._get_date("until")
+        
+    def _clean_commit(self, line):
+        cdata = line['data']
+        cleaned_line =  \
+        {
+            'repo': line['origin'],
+            'hash': cdata['commit'],
+            'author': cdata['Author'],
+            'category': "commit",
+            'created_date': self._str_to_dt_data(cdata['AuthorDate']),
+            'commit': cdata['Commit'],
+            'commit_date': self._str_to_dt_data(cdata['CommitDate']),
+            'files_no': len(cdata['files']),
+            'refs': cdata['refs'],
+            'parents': cdata['parents'],
+            'files': cdata['files']
+        }
+
+        actions = 0
+        for file in cdata['files']:
```

Comment:
If this is for removing "empty commits", those which touch no files, that's fine, But in the case of the metrics for commits, this would be a parameter to the specific metric (child of this one), as the specification tells: for example, in [the Git section of Code_Changes](https://github.com/chaoss/wg-evolution/blob/master/metrics/Code_Changes.md#git-parameters), "include merge commits" is a parameter of the metric... Which means that it should be one of the values uses for instantiating the Code_Changes metric, that would be a child of this root metric.

Reply:
Those few lines of code are a part of the the cleaning process of a raw commit. 
- **actions part**
This part adds an attribute called `files_action` to the new commit structure. This counts the number of files the commit affects. 

- **merge part**
This part adds the `merge` attribute to the cleaned commit, which is a boolean to determine if the commit is an actual commit or a merge commit.

Both of these are useful in the `compute` function for things like excluding merge commits or empty commits, just like you mentioned. 
The lines below are from the `compute` function, which would be part of any metric file which deals with commits.
```python
        if not empty:
            df = df[df['files_action'] != 0]
        if not merge:
            df = df[df['merge'] == False]
        return df['hash'].nunique()
```

Reply:
I think that's a good example of the kind of stuff that I would add in a child class for commits, not in the root.

Seeing now the whole picture, I think we're not that far apart. It is only that you were implementing in a single class what I had thought to be two classes: root (for the common functionality, which is just a bit more than reading JSON files) and the child for implementing a certain metric. Then, we can also learn that all the metrics related to commits, for example, have some common code, and maybe the final structure could be a root class, with classes for commits, issues, pull requests as children of it, and then specific metrics for commits, for example, being a child of the commits class.

Do you agree on this?